Source/InvocationCollection.cs

```diff
@@ -78,49 +78,120 @@ public int Count
 
 		public void Add(Invocation invocation)
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				this.invocations.Add(invocation);
+				if (this.count == this.capacity)
+				{
+					var targetCapacity = this.capacity == 0 ? 4 : (this.capacity * 2);
+					Array.Resize(ref this.invocations, targetCapacity);
+					this.capacity = targetCapacity;
+				}
+
+				this.invocations[this.count] = invocation;
+				this.count++;
 			}
 		}
 
 		public bool Any()
 		{
-			return this.invocations.Count > 0;
+			lock (this.invocationsLock)
+			{
+				return this.count > 0;
+			}
 		}
 
 		public void Clear()
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				this.invocations.Clear();
+				// Replace the collection so readers with a reference to the old collection aren't interrupted
+				this.invocations = null;
+				this.count = 0;
+				this.capacity = 0;
 			}
 		}
 
 		public Invocation[] ToArray()
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				return this.invocations.ToArray();
+				if (this.count == 0)
+				{
+					return new Invocation[0];
+				}
+
+				return this.invocations.Take(this.count).ToArray();
 			}
 		}
 
 		public Invocation[] ToArray(Func<Invocation, bool> predicate)
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				return this.invocations.Where(predicate).ToArray();
+				if (this.count == 0)
+				{
+					return new Invocation[0];
+				}
+
+				return this.invocations.Take(this.count).Where(predicate).ToArray();
```

Comment:
I wonder if you could use `Array.Copy` or `List<T>.CopyTo` instead of a LINQ query?

Reply:
I think it could be re-written to use those calls, but the predicate makes it a little harder. I'm not sure if there's a benefit to do so.

Reply:
Less unnecessary overhead, basically. This class is pretty low-level, I usually try to avoid LINQ in such places.

Reply:
(Sorry, I was an idiot above. I was looking at the wrong lines of code when I wrote the above, I completely overlooked the predicate.)